Create SOP for knative_up != 1

aliok opened the pull request:
Create a simple SOP for the case where `knative_up` metric is not `1`.

Users/services can create an alert on their own, using this metric and can use this document as a SOP for that alert.

cc @ReToCode 
cc @lberk 
cc @skonto 

@jewzaam is this a good SOP? Could please take a look?

@danielezonca is this usable for your project?

matzew commented on docs/sops/README.md:
```diff
@@ -0,0 +1,3 @@
+# OpenShift Serverless SOPS
```
SOPS: *S*tandard *O*perating *P*rocedure*s* ?

aliok replied:
updated

matzew commented on docs/sops/serverless-controller-knative-up.md:
````diff
@@ -0,0 +1,93 @@
+2. If they are not running, look at the pod's logs/events to see what may be causing the issues. Please make sure to grab the logs/events so they can be shared with the engineering team later:
+
+```bash
+# Get Pod names as we don't have a common label
````
should we introduce such a label? (wondering about this being scriptable)

aliok replied:
Created https://issues.redhat.com/browse/SRVCOM-2703

danielezonca commented on docs/sops/serverless-controller-knative-up.md:
```diff
@@ -0,0 +1,93 @@
+
+You should check both OpenShift Serverless operator and Knative Serving control plane status.
+
+## Steps
```
Let's add a "Prerequisite" sections, in this case I expect it is enough to have the cluster and a `oc` client logged in but it is a good template to follow

aliok replied:
I copy pasted the ModelServing SOP in the private SRE repo and used it as the template, but let me add a pre-reqs section. It is a good idea.

danielezonca commented on docs/sops/serverless-controller-knative-up.md:
```diff
@@ -0,0 +1,93 @@
+
+## Summary
+
+OpenShift Serverless does not create any alerts out-of-the box. However, users can create alerts based on the metrics exposed.
+
+The metric `knative_up{namespace="openshift-serverless", type="serving_status"}` should always have a value of `1`. 
```
I would personally create a simple list of alerts (in this repo?) that users can create so that here we can refer it also by name

aliok replied:
Good idea @danielezonca !

I've created a ticket for that now: https://issues.redhat.com/browse/SRVCOM-2707

I think that's a separate issue/work.

danielezonca commented on docs/sops/serverless-controller-knative-up.md:
````diff
@@ -0,0 +1,93 @@
+
+## Steps
+
+1. Check to see if all the pods are running in `openshift-serverless` namespace:
+
+```bash
+$ oc -n openshift-serverless get pods
+```
````
I would make this explicit: if the goal is to get all pods that are not running make the command already filter out only them like `--field-selector status.phase!=Running`

The reason of this comment is that this procedure should be as much as possible simple, for example if the command is going to return a lot of pods you might risk to don't see one of them not running

aliok replied:
I would prefer showing all pods.

Reason:
- Later on, there's an instruction to collect logs from pods. The `oc logs` command doesn't support something like "only logs of failing/unready/crashing pods" (it doesn't make sense to only collect the logs of those anyway. IMO, logs of the whole system should be collected)
- So, the person will collect the logs will see all logs, but they will see new pods that they haven't seen before.

We don't have too many pods anyway.

Also, the SOP I've used as a basis checks all logs, and not only the failing ones.

If one day and SRE comes and says we need to go this route, I can do it. I feel it is a bit of premature optimization at this point.

danielezonca commented on docs/sops/serverless-controller-knative-up.md:
````diff
@@ -0,0 +1,93 @@
+5. Check to see if all the pods are running in `knative-serving` namespace:
+
+```bash
+$ oc -n knative-serving get pods -l app.kubernetes.io/name=knative-serving
````
Same as above

danielezonca reviewed (commented):
Thanks @aliok for the PR, overall I think it is a good example of SOP, I have some comments but overall it looks good to me :)

matzew commented:
Is this generally something we want to "proper" document?  (e.g. on some "openshift docs" format)?

Or are we stashing these documents here and eventually move them else?

ReToCode commented on docs/sops/serverless-controller-knative-up.md:
````diff
@@ -0,0 +1,93 @@
+$ oc -n openshift-serverless get pods
+```
+
+2. If they are not running, look at the pod's logs/events to see what may be causing the issues. Please make sure to grab the logs/events so they can be shared with the engineering team later:
````
I'm not sure aware of the target audience for a SOP, but I'd assume that anyone operating the cluster already knows how to do to this? Would it be sufficient to just list which pods to take a look at, similar to 
https://knative.dev/docs/install/troubleshooting/

matzew replied:
I like that idea from upstream, instead of doing too much of an introduction, which can be assumed as known (I guess)

aliok replied:
I've followed the SOPs in SRE repository in managed tenants. I would prefer keeping this format. It is them who is going to use the SOPs.

ReToCode commented on docs/sops/serverless-controller-knative-up.md:
```diff
@@ -0,0 +1,93 @@
+# OpenShift Serverless Knative Serving Control Plane Readiness
```
This seems like it is targeting only the operator pods. So that is not really "Serving", right? 
For Serving we'd also need to look at all pods in `knative-serving`, `knative-serving-ingress` and (when istio is used) `istio-system` (or a custom namespace where istio is deployed).

aliok replied:
Down in the file, there are instructions to check if pods in `knative-serving` is running.

But, let me add the same thing for other namespaces.

Thanks

aliok replied:
Added some text for Ingress now.

Thanks @ReToCode for the suggestion.

ReToCode commented on docs/sops/serverless-controller-knative-up.md:
```diff
@@ -0,0 +1,93 @@
+
+This should result in new pods getting deployed, attempt step (5) again and see if the pods achieve running state.
+
+8. If the problem persists, capture the logs and escalate to OpenShift Serverless engineering team.
```
Yes, although a non-knative specific must-gather would help as well.

aliok replied:
The idea to escalate to the engineering team with the _logs_. 
Then if eng team asks for a must-gather after checking the collected logs, SREs can do it.

So, IMO, let's not add it here.

ReToCode reviewed (commented):
I like it. I'm also wondering about https://github.com/openshift-knative/serverless-operator/pull/2272#issuecomment-1727019166.

aliok commented on docs/sops/serverless-controller-knative-up.md:
```diff
@@ -0,0 +1,93 @@
+$ done
+
+
+# Check events 
+$ oc -n openshift-serverless get events
```
I do not have that information right now.
Once we have it under operation, we can add some known events gradually.

aliok commented:
@matzew  @ReToCode 

> Is this generally something we want to "proper" document? (e.g. on some "openshift docs" format)?
> Or are we stashing these documents here and eventually move them else?

The idea is that other teams would need this SOP in the future. I have tried pushing the same SOP in the SRE repository hoping this would help with collaboration, but it didn't.

Now, I am preparing a single SOP in advance, which will be needed. And once it is needed, it can be consumed pretty easily. We will need more SOPs, but that's something I need to work on with other teams.

> Or are we stashing these documents here and eventually move them else?
So, basically stashing these documents here and later move them somewhere else.

aliok commented on docs/sops/serverless-controller-knative-up.md:
````diff
@@ -0,0 +1,93 @@
+
+```bash
+# Get Pod names as we don't have a common label
+$ SERVERLESS_OPERATOR_POD_NAMES_STR=$(oc -n openshift-serverless get pods --template '{{range .items}}{{.metadata.name}}{{" "}}{{end}}')
````
Same as https://github.com/openshift-knative/serverless-operator/pull/2272#discussion_r1332718317

aliok commented:
@ReToCode @danielezonca 
Thanks very much for your reviews. I really appreciate them.

Can you have a new look please? I've addressed all of your comments (made code changes OR replied with my reasoning)

ReToCode reviewed (commented):
A thing I noted when checking the markdown view, adding `$` is not ideal if you want to copy the commands. Especially if they are multi-line. I think dropping that would help with UX here.

Other than that, LGTM.

aliok commented:
>  I think dropping that would help with UX here.

I've dropped the `$` prefix for multiline scripts now.

skonto commented:
Not sure if anyone mentioned this but `knative_up` is a gauge and users should be aware of what prom query they should use: https://www.robustperception.io/alerting-on-gauges-in-prometheus-2-0/.

skonto commented on docs/sops/serverless-controller-knative-up.md:
```diff
@@ -0,0 +1,149 @@
+
+OpenShift Serverless does not create any alerts out-of-the box. However, users can create alerts based on the metrics exposed.
+
+The metric `knative_up{namespace="openshift-serverless", type="serving_status"}` should always have a value of `1`. 
```
Actually for gauges it is better to measure it over some period of time eg. https://www.robustperception.io/alerting-on-gauges-in-prometheus-2-0/.

aliok replied:
Thanks for the great input @skonto !

`avg_over_time(up{job="jobname"}[5m]) < 0.9` we should use this kind of structure in https://issues.redhat.com/browse/SRVCOM-2704

aliok replied:
Also updated the text here to reflect that `knative_up` is a gauge and consumers should measure it over period of time. However, we don't provide alerts in this PR. That's part of SRVCOM-2704.

aliok commented on docs/sops/serverless-controller-knative-up.md:
```diff
@@ -0,0 +1,149 @@
+
+## Summary
+
+OpenShift Serverless does not create any alerts out-of-the box. However, users can create alerts based on the metrics exposed.
+
+The metric `knative_up{namespace="openshift-serverless", type="serving_status"}` should always have a value of `1`. 
```
```suggestion
OpenShift Serverless does not create any alerts out-of-the box. However, users can create alerts based on the gauges and metrics exposed.

The gauge `knative_up{namespace="openshift-serverless", type="serving_status"}` should always have a value of `1`.  You should create an alert as described in https://www.robustperception.io/alerting-on-gauges-in-prometheus-2-0/ for using this gauge and measure it over some period of time.
```

aliok commented:
/retest

skonto commented on docs/sops/serverless-controller-knative-up.md:
````diff
@@ -0,0 +1,149 @@
+    enabled: false
+```
+
+In this case, Istio is used as the Knative Serving Ingress.
````
Probably we should mention that we use ServiceMesh and people should also check their setup there for issues.

aliok replied:
I will make some adjustments to the text here.

aliok replied:
I mentioned ServiceMesh too now.

>  people should also check their setup there for issues
This is already written in step 11 below

skonto commented on docs/sops/serverless-controller-knative-up.md:
```diff
@@ -0,0 +1,149 @@
+
+11. If Istio is used as the Knative Serving Ingress, check Istio status. You may find the SOPs for Istio by contacting Istio support. 
+
+12. If the problem persists, capture the logs and escalate to OpenShift Serverless engineering team with a Knative ["must-gather"](https://github.com/openshift-knative/must-gather) dump.
```
Could you add the command here:

```oc adm must-gather --image=registry.redhat.io/openshift-serverless-1/svls-must-gather-rhel8 ```

That is the image I see people use. Not sure if there is a difference compared to quay.io/openshift-knative/must-gather.

Also it would be useful to add the SM one.

skonto replied:
I think SM uses `registry.redhat.io/openshift-service-mesh/istio-must-gather-rhel8:2.4.0` but needs verification.

aliok replied:
Removed the mention to must gather.

skonto replied:
They could also reach support or some contact?

aliok replied:
Good feedback, thanks. Let me find some links to put in here :)

aliok replied:
@skonto added some links about how to reach out to SM team.
(I've asked in Istio team channel)

skonto commented on docs/sops/serverless-controller-knative-up.md:
````diff
@@ -0,0 +1,149 @@
+$ oc -n openshift-serverless get pods
+```
+
+2. If they are not running, look at the pod's logs/events to see what may be causing the issues. Please make sure to grab the logs/events so they can be shared with the engineering team later:
````
Maybe for events and particular pods this is more focused:
https://stackoverflow.com/questions/51931113/kubectl-get-events-only-for-a-pod

aliok replied:
I followed existing SOPs and use them as template.

I think it is a good idea to keep this instruction more generic. We already give instructions in the sections above for collecting events and logs.

skonto replied:
Ok sure.

skonto commented on docs/sops/serverless-controller-knative-up.md:
```diff
@@ -0,0 +1,149 @@
+# OpenShift Serverless Knative Serving Control Plane Readiness
```
Are you planning to add anything for Eventing too?

aliok replied:
Not right now. This SOP is needed for RHODS, but they don't use Eventing.

ReToCode reviewed (approved):
/lgtm

aliok commented:
/unhold
